engine/apps/alerts/tasks/check_escalation_finished.py

```diff
@@ -1,48 +1,130 @@
+import datetime
+import typing
+
+import requests
+from celery import shared_task
 from django.apps import apps
 from django.conf import settings
 from django.db.models import Q
 from django.utils import timezone
 
 from apps.alerts.tasks.task_logger import task_logger
-from common.custom_celery_tasks import shared_dedicated_queue_retry_task
+from common.database import get_random_readonly_database_key_if_present_otherwise_default
 
+if typing.TYPE_CHECKING:
+    from apps.alerts.models.alert_group import AlertGroup
 
-@shared_dedicated_queue_retry_task(
-    autoretry_for=(Exception,), retry_backoff=True, max_retries=1 if settings.DEBUG else None, default_retry_delay=60
-)
-def check_escalation_finished_task():
+
+class AlertGroupEscalationPolicyExecutionAuditException(BaseException):
+    """This exception is raised when an alert group's escalation policy did not execute execute properly for some reason"""
+
+
+def send_alert_group_escalation_auditor_task_heartbeat() -> None:
+    heartbeat_url = settings.ALERT_GROUP_ESCALATION_AUDITOR_CELERY_TASK_HEARTBEAT_URL
+    if heartbeat_url:
+        task_logger.info(f"Sending heartbeat to configured URL: {heartbeat_url}")
+        requests.get(heartbeat_url).raise_for_status()
+        task_logger.info(f"Heartbeat successfully sent to {heartbeat_url}")
+    else:
+        task_logger.info(f"Skipping sending heartbeat as no heartbeat URL is configured")
+
+
+def audit_alert_group_escalation(alert_group: "AlertGroup") -> None:
+    escalation_snapshot = alert_group.escalation_snapshot
+    alert_group_id = alert_group.id
+    base_msg = f"Alert group {alert_group_id}"
+
+    if not escalation_snapshot:
+        raise AlertGroupEscalationPolicyExecutionAuditException(
+            f"{base_msg} does not have an escalation snapshot associated with it, this should never occur"
+        )
+    task_logger.info(f"{base_msg} has an escalation snapshot associated with it, auditing if it executed properly")
+
+    escalation_policies_snapshots = escalation_snapshot.escalation_policies_snapshots
+
+    if not escalation_policies_snapshots:
+        # TODO: in what case is this valid?
+        task_logger.info(
+            f"{base_msg}'s escalation snapshot has an empty escalation_policies_snapshots, skipping further validation"
+        )
+        return
+    task_logger.info(
+        f"{base_msg}'s escalation snapshot has a populated escalation_policies_snapshots, continuing validation"
+    )
+
+    if escalation_snapshot.next_step_eta_is_valid() is False:
+        # TODO: is a next_step_eta value of None actually valid here, or should we raise an error?
```

Comment:
question here

Reply:
`next_step_eta` could be `None` if there are no escalation steps in the chain or escalation was not started. After executing the first escalation step, this field is set to some time value, depending on the step type. I'm not sure if there is any other way to make this happen.
But since you check above if escalation policies exist, I think, any other cases of `next_step_eta is None` should be invalid here 🤔

Reply:
good point. I'll raise an exception here 👍